internal/cluster/transition.go

```diff
@@ -142,33 +145,146 @@ func (th *transitionHandler) PostUpdateInstallationProgress(sw stateswitch.State
+	params, ok := args.(*TransitionArgsRefreshCluster)
 	if !ok {
 		return errors.New("PostCompleteInstallation invalid argument")
 	}
 
-	return th.updateTransitionCluster(logutil.FromContext(params.ctx, th.log), th.db, sCluster, params.reason)
+	status, reason := th.getClusterMonitoringOperatorsStatus(sCluster.cluster)
+
+	return th.updateCompletion(params.ctx, params.ocmClient, params.metricApi, params.eventHandler, sCluster, status == models.OperatorStatusAvailable, reason)
 }
 
-func (th *transitionHandler) isSuccess(stateSwitch stateswitch.StateSwitch, args stateswitch.TransitionArgs) (b bool, err error) {
-	params, _ := args.(*TransitionArgsCompleteInstallation)
-	return params.isSuccess, nil
+func (th *transitionHandler) updateCompletion(ctx context.Context, ocmClient *ocm.Client, metricAPI metrics.API, eventsHandler events.Handler,
+	sCluster *stateCluster, successfullyFinished bool, reason string) error {
+	log := logutil.FromContext(ctx, th.log)
+	cluster := sCluster.cluster
+
+	if successfullyFinished && ocmClient != nil && ocmClient.Config.WithAMSSubscriptions {
+		if err := ocmClient.AccountsMgmt.UpdateSubscriptionPostInstallation(ctx, cluster.AmsSubscriptionID, cluster.OpenshiftClusterID); err != nil {
+			err = errors.Wrapf(err, "Failed to update AMS subscription for cluster %s post installation", *cluster.ID)
+			log.Error(err)
+			return err
+		}
+	}
+
+	if err := th.updateTransitionCluster(log, th.db, sCluster, reason); err != nil {
+		err = errors.Wrapf(err, "Failed to update cluster %s completion in db", *cluster.ID)
+		log.Error(err)
+		return err
+	}
+
+	result := models.ClusterStatusInstalled
+
+	severity := models.EventSeverityInfo
+	eventMsg := fmt.Sprintf("Successfully finished installing cluster %s", cluster.Name)
+	if !successfullyFinished {
+		result = models.ClusterStatusError
+		severity = models.EventSeverityCritical
+		eventMsg = fmt.Sprintf("Failed installing cluster %s. Reason: %s", cluster.Name, reason)
+	}
+
+	metricAPI.ClusterInstallationFinished(log, result, cluster.OpenshiftVersion,
+		*cluster.ID, cluster.EmailDomain, cluster.InstallStartedAt)
+	eventsHandler.AddEvent(ctx, *cluster.ID, nil, severity, eventMsg, time.Now())
+
+	return nil
+}
+
+func (th *transitionHandler) isCompletedSuccess(stateSwitch stateswitch.StateSwitch, args stateswitch.TransitionArgs) (hasClusterComplete bool, err error) {
+	hasClusterComplete, status, err := th.hasClusterCompleteInstallation(stateSwitch, args)
+	if !hasClusterComplete {
+		return false, err
+	}
+
+	if status == nil {
+		return true, nil
+	}
+
+	return *status == models.OperatorStatusAvailable, nil
 }
 
-func (th *transitionHandler) notSuccess(stateSwitch stateswitch.StateSwitch, args stateswitch.TransitionArgs) (b bool, err error) {
-	params, _ := args.(*TransitionArgsCompleteInstallation)
-	return !params.isSuccess, nil
+func (th *transitionHandler) isCompletedFailure(stateSwitch stateswitch.StateSwitch, args stateswitch.TransitionArgs) (b bool, err error) {
```

Comment:
why we have 2 isCompletedSuccess and isCompletedFailure ?

Reply:
one for the transition to Installed state and onr yo the Error state